library/Booster/Builtin.hs

```diff
@@ -32,12 +33,83 @@ hooks :: Map ByteString BuiltinFunction
 hooks =
     Map.unions
         [ builtinsKEQUAL
+        , builtinsMAP
         ]
 
 ------------------------------------------------------------
 (~~>) :: ByteString -> BuiltinFunction -> (ByteString, BuiltinFunction)
 (~~>) = (,)
 
+------------------------------------------------------------
+-- MAP hooks
+-- only lookups and in_keys are implemented
+builtinsMAP :: Map ByteString BuiltinFunction
+builtinsMAP =
+    Map.mapKeys ("MAP." <>) $
+        Map.fromList
+            [ "lookup" ~~> mapLookupHook
+            , "lookupOrDefault" ~~> mapLookupOrDefaultHook
+            , "in_keys" ~~> mapInKeysHook
+            ]
+
+mapLookupHook :: BuiltinFunction
+mapLookupHook args
+    | [KMap _ pairs _mbRest, key] <- args =
```

Comment:
Will we eventually have to return something like `#bottom` here in case we know this is a fully concrete map?

Reply:
Probably that would be the correct way - we know the result is "undefined" so no point trying any equations.
 Our types currently do not allow `Term`s to be  `#Bottom`  - we would need to add an explicit term-bottom constructor.